### ISSUE TITLE
Cache should have subdirectories for e.g formulae and versions

### ISSUE THREAD

**Issue by MikeMcQuaid** [Homebrew/brew#568]
The `HOMEBREW_CACHE` should be reordered to have a format something like the following:

```
formula/version/source/file.tgz
formula/version/bottles/bottle.tar.gz
resources/resource-$SHA256.tar.gz
patches/patch.diff
```

This will allow simplifying the cleanup of the cache without needing to know as much about formula state.

**Comment by chdiza** [comment]
Is there a reason why the source tarball go in a subdir named "source", rather than sitting loose in "version"?  I'm not keen on subdirs that only hold one item, which seems to be what the "source" dir would be.  (I've never seen multiple sourceballs for one version of a formula, AFAICR.)

"Resources" and "patches" are fine because there will frequently be multiple children of each if someone brews enough formulas.

"Bottles" will only ever have one child except for bottle revisions, which it seems to me is akin to "resources" and "patches".

**Comment by MikeMcQuaid** [comment]
No real reason, no. I'm not that bothered on the exact approach and you've made good points.

**Comment by chdiza** [comment]
> `resources/resource-$SHA256.tar.gz`

As I understand this, we would get for example:
```
resources/git--html-SOMESHA256.tar.gz
```
But then we lose track of what the version is (if `resources` here doesn't live under `formula/version`).  I think we'd need:
```
resources/git--html-2.9.1-SOMESHA256.tar.gz
```

**Comment by MikeMcQuaid** [comment]
> But then we lose track of what the version is (if resources here doesn't live under formula/version). I think we'd need:

We specifically don't need/want the version. Alternatively, we could just always use the original filename instead.

**Comment by UniqMartin** [comment]
If we keep prefixing resources with specific formula names and continue to associate them with specific formulae (though not specific versions of that formula), then it would make sense to save them in `<formula>/resources/[<resource-name>-]<resource-sha256>.tar.gz`.

If we would like to move to a model where we share resources across all formulae, e.g. based on their origin or SHA256, we could put them in a top-level `resources` directory. The upside is that this will avoid duplicating some resources that are used in multiple formulae. The downside I see here is that it would become more difficult to tell which resource belongs to which formula (e.g. for the sake of cleaning the cache). There are also subtle issues this can cause, e.g. what happens if two formulae are installed in parallel and both require the same resource. Do we need more sophisticated locking? Do we prevent parallel build/fetch entirely?

**Comment by MikeMcQuaid** [comment]
@UniqMartin Another option is a combined approach where we use hardlinks to reduce space.

**Comment by UniqMartin** [comment]
I'm personally not a big fan of hard links, as they make things quite intransparent and I'm a bit scared of the consequences that we (or someone else) might end up modifying a file in place.

Would it maybe make sense to adopt ideas from Git LFS for our cache structure, i.e. all files are stored in a shared directory and the name is derived from the used hash function and the hash? Everything else would be stored in formula-specific directories, but is just a symlink into this shared directory, e.g.

```
-- CACHE
    |-- .blob
    |   \-- sha256
    |       |-- 61
    |       |   \-- d1a4e3bdc1846267c984b478e3a5a822a8a14a1169153670a9a61402ab1d40
    |       \-- dd
    |           \-- 9c3e37c266e4fefba68e444e2a00538b3c902dd31bf4912d90dca6d830a2a1
    \-- binutils # <- Formula#name
         \-- 2.26.1 # <- Formula#pkg_version
             |-- bottle-el_capitan.tar.gz -> ../../.blob/sha256/61/d1a4e3bdc1846267c984b478e3a5a822a8a14a1169153670a9a61402ab1d40
             |-- patch-NAME.EXT -> ../../.blob/sha256/…
             |-- resource-NAME.EXT -> ../../.blob/sha256/…
             \-- source.tar.gz -> ../../.blob/sha256/dd/9c3e37c266e4fefba68e444e2a00538b3c902dd31bf4912d90dca6d830a2a1
```

Populating the cache would be a matter of checking whether a blob with the given hash exists and setting up a symbolic link if it does. If it does not, create a lock on the hash (to avoid parallel download of the same blob) and download it. Cleaning up the cache would become a two-step process: First identify all symbolic links that can be removed and do so, then remove all blobs not referenced by any symbolic link in the cache repository. (The second step should be really cheap because it's just a file system traversal not involving any formula loading.)

Because duplicate blobs are now really cheap, we could force *all* things for a specific formula version into a `<formula>/<pkg_version>` directory (as hinted at in the tree above). This would make it extremely easy to figure out the corresponding formula or formula version and cleaning the cache for a given formula would be as simple as recursively wiping a certain directory.

Not sure if this is indeed a sane idea. Just brainstorming here …

**Comment by chdiza** [comment]
Regarding the directory structure, I think that source tarballs and patches should be easy to find in the cache, and not obscured by links or excessively-nested dirs or opaquely-named dirs.  Sometimes manual deletion from the cache is necessary, and sometimes manual insertion is convenient.  This will become very difficult if we have an overly complex structure or obscure file names.

Example case: suppose I want to call `brew cleanup` to do some housekeeping, but I also want to be certain that I don't accidentally delete the source for `foo` because I know the server hosting `foo` has been flakey.  I thus want to go to the cache and make a copy of the source tarball and put it outside the cache for safekeeping.  It shouldn't be a grand project to find the source tarball in the cache.  That tarball can be a few levels deep, but it should be clear how to get there.  E.g.,  `foo/1.0/source/stable/`.  The user should definitely not have to open the formula for foo, copy the checksum, and go hunting in the cache for files matching that checksum.

**Comment by UniqMartin** [comment]
> Example case: suppose I want to call brew cleanup to do some housekeeping, but I also want to be certain that I don't accidentally delete the source for foo because I know the server hosting foo has been flakey. I thus want to go to the cache and make a copy of the source tarball and put it outside the cache for safekeeping. It shouldn't be a grand project to find the source tarball in the cache.

Assuming we went with my suggestion, this case is covered and it would be trivial to do this, e.g. `cd "$(brew --cache)/foo/1.0" && cp * /path/to/safe/location/`. Symbolic links are resolved to the file they are pointing to by `cp` (unless you explicitly tell it to behave differently), thus you'd get a copy of everything needed by version 1.0 of `foo` as regular files, including patches/resources that might be on the same flakey server.

**Comment by chdiza** [comment]
But that forces the user to the terminal, and to *know* that fact about `cp`.  For users who don't know all that stuff, I think they should still be able to easily navigate to the source tarball, even if it's entirely within the Finder.

**Comment by UniqMartin** [comment]
The user you're describing doesn't sound like a user targeted by Homebrew. But even if that was the case, the use case you're describing is definitely not that of a novice user and thus it's fair to assume they are comfortable with some level of command line usage. I suspect most users (for better or worse) aren't even aware of the cache and don't think about manipulating it directly.

**Comment by chdiza** [comment]
Don't get hung up on the example :) The point is that manual extraction from and insertion into the cache should be as easy as possible, for all users, including the kind of newbs to the commandline that HB bends over backwards to accommodate.

**Comment by MikeMcQuaid** [comment]
@UniqMartin I like the idea but it feels very complicated to me; the goal of this change would be to make things less rather than more complicated. I'm happy for resource deduplication to be done in a second pass; I'm not convinced it's nearly as important.

**Comment by MikeMcQuaid** [comment]
> The point is that manual extraction from and insertion into the cache should be as easy as possible, for all users, including the kind of newbs to the commandline that HB bends over backwards to accommodate.

Agreed 👍

**Comment by UniqMartin** [comment]
> The point is that manual extraction from and insertion into the cache should be as easy as possible, for all users, […]

I disagree. To me this sounds like an implementation detail that we should be free to change at any time and not like something average users should be messing with.

> […], including the kind of newbs to the commandline that HB bends over backwards to accommodate.

We accommodate those users on the surface, by hopefully making our interface as easy to understand and explore as possible. If easy messing with implementation details was ever an objective, why am I not aware of any web browser where I can just go into the cache directory (if I even manage to find it) and then can easily cherry pick some files from there?

Maybe I'm just missing the point of accommodating a very slim minority while potentially all users could benefit from a redesign of the cache that is geared towards being more machine readable. (And I'm not claiming that my idea will achieve this goal. It's just an idea after all.)

**Comment by MikeMcQuaid** [comment]
@UniqMartin I don't think the filesystem should be designed to be machine readable rather than human readable. Homebrew has stated that the filesystem (and command stdout) are our interface and I think favouring simplicity is wise as it makes it easier to interoperate with other e.g. Unix tools. In this case a simple solution will be easier to implement and maintain. As mentioned, we can investigate more complicated deduplication later (for which I still favour hard links) but this issue is relatively tightly scoped to a new cache subdirectory format.

**Comment by joshka** [comment]
I'm just about to get started on implementing this. Unless there's objections, I plan to skip the deduplication and linking on the first pass and will add another issue to continue discussion later.

Here's what I think makes sense:

```
Cache/<formula-name>/<formula-version>/<formula-name>-<formula-version>.source.tar.gz
Cache/<formula-name>/<formula-version>/<formula-name>-<formula-version>.bottle-<osx-codename>.tar.gz
Cache/<formula-name>/<formula-version>/<formula-name>-<formula-version>.resource-<resource-name>.tar.gz
Cache/<formula-name>/<formula-version>/<formula-name>-<formula-version>.patch-<patch-name>.tar.gz
```

I realize that this repeats the formula name and version in the filename. This is intentional as it makes it easier for users to identify large files and make sensible decisions on keeping or not for manual cache management tasks from finder compared to filenames like source.tgz. I'm hoping it satisfies @chdiza's make it simple radar as well.

Note, this drops the resource version from resource filenames as mentioned in #517, and I'll intentionally only use the url basename / non version specific filename.

**Comment by ilovezfs** [comment]
IMO dedup should be part of the initial design, not bolted on later.

**Comment by joshka** [comment]
Ok then, some questions with my suggested answers.

1. Q: De-dupe all files or just resources?
A: Both as formula version may be bumped while pointing to the same version of a source.

2. Pros and cons of sym-link vs hard-link?
A: I think @MikeMcQuaid likely suggested hard-links as it makes the distinction of whether the file is a normal file or symlink immaterial for the scenario where you want to copy files into or out of the cache. @UniqMartin mentioned modification in place, but I'm not sure that I can see how sym-links instead of hard-links would solve this. Regardless, we always verify checksums prior to install.

3. Q: Store de-duplicated files under which folder?
A: `Cache/by-sha256/<2-char-prefix>/<rest-of-filename>`. 2 character prefix to ensure this directory doesn't get too large. This matches git's `.git/objects` layout. This naming scheme doesn't need to be particularly user friendly if we're using hard-links because the user friendly file (`/formula/ver/file.ext`) is an equivalent to these files in every way. It only exists to make de-duplication work.

4. Q: Cleanup logic?
A: Any file under `Cache/by-sha256` pointing to a single inode will be deleted after the main cleanup happens.

Another option that augments the hard link approach, is that non hard-linked files could continue to just work in the cache. We can perform a deduplication step during initial verification of the file's checksum, or manually (perhaps a dev-cmd?). This avoids the need to find / sort out hashes / folder structure etc. while copying. This is the main reason that I think we don't need to do de-duplication at the same time as folder organization. "Bolting on" a backwards compatible approach seems fine to me.

**Comment by ilovezfs** [comment]
I don't think you need a special layout to dedup. hardlinks can be between the files in the usual subdirectories. Also, dedup shouldn't be a second step, since we should avoid duplicate downloads wherever possible.

**Comment by joshka** [comment]
Where do you store the checksum though?
Let's say you install a formula, uninstall it, brew update then install the next version of the formula. The source may have not changed (i.e. a formula fixup), but we don't necessarily have the ability to avoid the duplicate download of a file that we no longer have a checksum lying around for.

Additionally, if it's in the usual subdirectory, we end up having to search all directories for files matching.

**Comment by ilovezfs** [comment]
We still have to formula the prior file in the .brew directory, so you can probably check the file from the just-prior-version. I don't think we need to worry about dedup where the file was from a version-before-the-prior-version or from some other formula.

**Comment by joshka** [comment]
What's the `.brew` directory?

Define 'just-prior' ;) I'm not sure you can hit that requirement easily. (side-note: the folder scheme above doesn't take into account version schemes, should it?).

Have you done any checks to get a feel for which patterns of deduplication would be useful? @UniqMartin mentioned resources shared between formulae. Is this hypothetical or a real world need?

**Comment by ilovezfs** [comment]
```
$ find /usr/local/Cellar -regex '.*\.brew.*' -type f|head -10
/usr/local/Cellar/ack/2.16/.brew/ack.rb
/usr/local/Cellar/ack/2.18/.brew/ack.rb
/usr/local/Cellar/bash/4.4.12/.brew/bash.rb
/usr/local/Cellar/bash-completion/1.3_2/.brew/bash-completion.rb
/usr/local/Cellar/cmake/3.7.2/.brew/cmake.rb
/usr/local/Cellar/coreutils/8.27/.brew/coreutils.rb
/usr/local/Cellar/curl/7.53.1/.brew/curl.rb
/usr/local/Cellar/direnv/2.11.3/.brew/direnv.rb
/usr/local/Cellar/dos2unix/7.3.4/.brew/dos2unix.rb
/usr/local/Cellar/doxygen/1.8.13_1/.brew/doxygen.rb
```

Just prior meaning whichever installed keg has the greatest version less than the given version.

Which folder scheme?

>resources shared between formulae

Python formulae probably have the most duplication between formulae, but the pypi tarballs tend to be relatively small anyway.

**Comment by joshka** [comment]
> `$ find /usr/local/Cellar -regex '.*\.brew.*' -type f|head -10`

A counterpoint is that uninstalling removes the directory.

```
$ brew install ag
==> Downloading https://homebrew.bintray.com/bottles/the_silver_searcher-1.0.2.sierra.bottle.tar.gz
Already downloaded: /Users/joshka/Library/Caches/Homebrew/the_silver_searcher-1.0.2.sierra.bottle.tar.gz
==> Pouring the_silver_searcher-1.0.2.sierra.bottle.tar.gz
==> Caveats
...
==> Summary
🍺  /usr/local/Cellar/the_silver_searcher/1.0.2: 10 files, 113.1KB

$ brew uninstall ag
Uninstalling /usr/local/Cellar/the_silver_searcher/1.0.2... (10 files, 113.1KB)

$ ls /usr/local/Cellar/ag/
ls: /usr/local/Cellar/ag/: No such file or directory
```

>Which folder scheme?

https://github.com/Homebrew/brew/issues/568#issuecomment-289169115

**Comment by ilovezfs** [comment]
>A counterpoint is that uninstalling removes the directory.

You could always copy the rb file (or, equivalently, some more convenient distillation of its checksums) into the cache as well if dedup against files of uninstalled kegs is a design goal.

>#568 (comment)

Yes it probably needs to account for version scheme, revision, and rebuild since brew versions are basically tuples of (tap, formula, version_scheme, version, revision, rebuild)

**Comment by joshka** [comment]
>You could always copy the rb file (or, equivalently, some more convenient distillation of its checksums) into the cache as well if dedup against files of uninstalled kegs is a design goal.

That seems like it would add more complexity to the homebrew implementation than a hard link.
What's the major de-dupe goal? What's the major de-dupe scenario that you saw as being critical enough to this issue that it blocks implementing as originally scoped (only fixing folders). I'm guessing it's the version bump that you'd hit when upgrading?

> Yes it probably needs to account for version scheme, revision, and rebuild since brew versions are basically tuples of (tap, formula, version_scheme, version, revision, rebuild)

The cache ends up quite deep with all that info:
`Cache/homebrew/homebrew-core/<formula-name>/<version-scheme>/<version>/<revision>/<rebuild>/source.tgz`. I'm assuming we collapse a part if it's omitted? You'll notice that working out which of those folders is the "previous version" then becomes more complex as well.

**Comment by joshka** [comment]
> You could always copy the rb file (or, equivalently, some more convenient distillation of its checksums) 

P.s. I often do this for Windows installers delivered as cd images using a sidecar .sha1 / .sha256 file that can be checked with `shasum`. Makes faulty usb copies easier to detect and mitigate.

**Comment by ilovezfs** [comment]
I'm still suggesting hard links. I'm not proposing that you keep nesting folders to account for version_scheme etc., only that the implementation will likely need to be aware of them.

I wondering if some variation of Martin's layout using hardlinks would be easier to manage while still being human readable. Thinking out loud, the checksum could be part of the file name so that identifying which blob to delete manually wouldn't be mysterious or require a specialized tool.

Assuming the "bar" resource is unchanged between 1.2.3, 1.2.4, and 1.2.4 revision 1:
```
foo/1.2.3/resource-bar-6e41057c8f50f661d800099471f769209480efa53b8a886969d7ec6db60a2208.tar.gz
foo/1.2.4/resource-bar-6e41057c8f50f661d800099471f769209480efa53b8a886969d7ec6db60a2208.tar.gz
foo/1.2.4_1/resource-bar-6e41057c8f50f661d800099471f769209480efa53b8a886969d7ec6db60a2208.tar.gz
blob/6e41057c8f50f661d800099471f769209480efa53b8a886969d7ec6db60a2208.tar.gz
```

**Comment by MikeMcQuaid** [comment]
> Cache/<formula-name>/<formula-version>/<formula-name>-<formula-version>.resource-<resource-name>.tar.gz
> Cache/<formula-name>/<formula-version>/<formula-name>-<formula-version>.patch-<patch-name>.tar.gz

I don't think these two should have the formula version in them as they don't necessarily change from version to version. 

> IMO dedup should be part of the initial design, not bolted on later.

I think it should be part of the design but I don't think we need it in v1 of the new layout.

> Both as formula version may be bumped while pointing to the same version of a source.

I don't understand this.

> Store de-duplicated files under which folder?

If you use hard links you can just duplicate the files in each location and let the filesystem handle it.

> I don't think you need a special layout to dedup. Hardlinks can be between the files in the usual subdirectories. Also, dedup shouldn't be a second step, since we should avoid duplicate downloads wherever possible.

Agreed.

> Is this hypothetical or a real world need?

It's not a need but it's a nice-to-have and is a good way to test de-dupe easily.

> foo/1.2.3/resource-bar-6e41057c8f50f661d800099471f769209480efa53b8a886969d7ec6db60a2208.tar.gz

I'd love to avoid the checksum in the name if it's technically possible. If we do really want to include it using a short, Git-style hash (e.g. 4-6 characters) would be good.

> foo/1.2.4_1/resource-bar-6e41057c8f50f661d800099471f769209480efa53b8a886969d7ec6db60a2208.tar.gz

I don't think the cache should be storing things based on revisions, personally, if possible.

**Comment by MikeMcQuaid** [comment]
I no longer care about this.
